### PR TITLE
Feature/receipts in state tree

### PR DESCRIPTION
**Pull request by liayoo**
- Record tx receipts in the state tree
  - `/receipts/<tx_hash>`: { address, block_number, gas_amount, gas_cost, result }
- Remove old tx receipts from the state tree (i.e. receipts from 1000 blocks ago)

#493

### CODE REVIEW

File: db/index.js [hunk 1]
```diff
@@ -1083,6 +1088,48 @@ class DB {
...
+     } else {
+       sanitized.code = result.code;
+      if (result.error_message) {
+        sanitized.error_message = result.error_message;
+      }
+    }
+     return sanitized;
+  }
+
+  recordReceipt(auth, timestamp, tx, blockNumber, result) {
```
Comment (platfowner) [inline db/index.js:1111]: minor: what about more specific param name e.g. 'txExecResult'?

File: db/index.js [hunk 2]
```diff
@@ -1083,6 +1088,48 @@ class DB {
     return true;
   }
 
+  static sanitizeResult(result) {
```
Comment (platfowner) [inline db/index.js:1091]: What about more specific name e.g. sanitizeTxExecResult()?
If this method does more than just normalizing the result, what about other names like simplifyTxExecResult() or something?

File: db/index.js [hunk 3]
```diff
@@ -1083,6 +1088,48 @@ class DB {
...
+  recordReceipt(auth, timestamp, tx, blockNumber, result) {
+    const receiptPath = PathUtil.getReceiptsPath(tx.hash);
+    const gasAmountSum = _.get(result, 'gas_amount_total.bandwidth.service', 0) +
+        _.get(result, 'gas_amount_total.state.service', 0);
+    const receipt = {
+      address: auth.addr,
+      block_number: blockNumber,
+      gas_amount: gasAmountSum,
+      gas_cost: result.gas_cost_total,
+      result: DB.sanitizeResult(result)
```
Comment (platfowner) [inline db/index.js:1120]: what about 'tx_exec_result' or just 'exec_result'?

File: db/index.js [hunk 4]
```diff
@@ -1083,6 +1088,48 @@ class DB {
...
+  }
+
+  recordReceipt(auth, timestamp, tx, blockNumber, result) {
+    const receiptPath = PathUtil.getReceiptsPath(tx.hash);
+    const gasAmountSum = _.get(result, 'gas_amount_total.bandwidth.service', 0) +
+        _.get(result, 'gas_amount_total.state.service', 0);
+    const receipt = {
+      address: auth.addr,
+      block_number: blockNumber,
+      gas_amount: gasAmountSum,
```
Comment (platfowner) [inline db/index.js:1118]: Would it be helpful for users if it logs the detailed (or raw) information?
e.g. `gas_amount: _.get(result, 'gas_amount_total', 0),`?

Comment (platfowner) [reply db/index.js:1118]: As we're not allowing empty object setting now, I guess the empty value setting part (e.g. apps: {}) needs to be updated (see https://github.com/ainblockchain/ain-blockchain/blob/develop/db/state-util.js#L140 ,
https://github.com/ainblockchain/ain-blockchain/blob/develop/db/index.js#L917 , and others).

Comment (platfowner) [reply db/index.js:1118]: Just leaving the detailed part of the gas amount in txExecResult and renaming 'gas_amount' -> 'gas_amount_by_transfer' here would be a better option?

File: db/index.js [hunk 5]
```diff
@@ -1083,6 +1088,48 @@ class DB {
...
+    }
+     return sanitized;
+  }
+
+  recordReceipt(auth, timestamp, tx, blockNumber, result) {
+    const receiptPath = PathUtil.getReceiptsPath(tx.hash);
+    const gasAmountSum = _.get(result, 'gas_amount_total.bandwidth.service', 0) +
+        _.get(result, 'gas_amount_total.state.service', 0);
+    const receipt = {
+      address: auth.addr,
```
Comment (platfowner) [inline db/index.js:1116]: Could you add these properties to PredefinedDbPaths as they're state paths (including 'billing' below)?

File: db/index.js (hunk 4 above)
Comment (platfowner) [reply db/index.js:1118]: Would 'bandwidth_gas_amount' be a better name?

Comment (platfowner) [reply db/index.js:1118]: Ah, sorry for now 'gas_amount' looks just fine to me.

File: db/index.js (hunk 1 above)
Comment (liayoo) [reply db/index.js]: Done, thanks!

File: db/index.js (hunk 2 above)
Comment (liayoo) [reply db/index.js]: Removed the sanitization logic as discussed offline :)

File: db/index.js (hunk 3 above)
Comment (liayoo) [reply db/index.js]: Updated to exec_result. Thanks for the suggestion!

File: db/index.js (hunk 4 above)
Comment (liayoo) [reply db/index.js]: Added the raw exec_result field which includes all the result code, error messages, gas amounts & costs of the operations.

File: db/index.js (hunk 5 above)
Comment (liayoo) [reply db/index.js]: Done!

**Comment by liayoo** [comment]
Addressed all comments, PTAL @platfowner

File: db/functions.js [hunk 6]
```diff
@@ -427,10 +423,10 @@ class Functions {
   }
 
   returnFuncResult(context, code, extraGasAmount = 0) {
-    const opResultList = Functions.getOpResultList(context);
+    const opResultListObj = CommonUtil.convertListToObj(context.opResultList);
     const funcResultToReturn = {};
-    if (!CommonUtil.isEmpty(opResultList)) {
-      funcResultToReturn.op_results = opResultList;
+    if (CommonUtil.isDict(opResultListObj)) {
```
Comment (platfowner) [inline db/functions.js:428]: question: didn't isEmpty() work? I think isEmpty() case includes null.

File: client/index.js [hunk 7]
```diff
@@ -398,6 +398,22 @@ app.get('/get_transaction', (req, res, next) => {
     .end();
 });
 
+app.get('/get_block_by_hash', (req, res, next) => {
```
Comment (platfowner) [inline client/index.js:401]: suggestion: what about adding json-rpc APIs as well?

File: common/constants.js [hunk 8]
```diff
@@ -44,6 +44,8 @@ const PORT = process.env.PORT || getPortNumber(8080, 8080);
 const P2P_PORT = process.env.P2P_PORT || getPortNumber(5000, 5000);
 const LIGHTWEIGHT = CommonUtil.convertEnvVarInputToBool(process.env.LIGHTWEIGHT);
 const SYNC_MODE = process.env.SYNC_MODE || 'full';
+const MAX_NUM_BLOCKS_RECEIPTS = process.env.MAX_NUM_BLOCKS_RECEIPTS ?
```
Comment (platfowner) [inline common/constants.js:47]: Could you consider simpler name MAX_NUM_RECEIPTS as neither the receipts nor this limit is a per-block value?

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments.
Thanks!

File: common/constants.js (hunk 8 above)
Comment (liayoo) [reply common/constants.js:47]: It's actually not the max number of receipts we can keep, but the max number of blocks that we can keep the receipts in the state for. Maybe `MAX_NUM_BLOCKS_FOR_RECEIPTS_IN_STATE` makes more sense? Please let me know if you have a better idea!

Comment (platfowner) [reply common/constants.js:47]: I see. That makes sense.
Then what about `MAX_BLOCK_NUMBERS_FOR_RECEIPTS`?

File: client/index.js (hunk 7 above)
Comment (liayoo) [reply client/index.js:401]: Thanks for the suggestion! Checked that we already have ain_getBlockByHash and ain_getBlockByNumber :)

File: db/functions.js (hunk 6 above)
Comment (liayoo) [reply db/functions.js]: Right, thanks for pointing that out. Reverted!

File: common/constants.js (hunk 8 above)
Comment (liayoo) [reply common/constants.js]: Sounds good to me! Thanks for the idea!

**Comment by liayoo** [comment]
Thanks for the quick review & for sharing ideas with me! Now merging..

**Pull request merged**